### PR TITLE
Host irs: add support for `MatmulOp`

### PR DESCRIPTION
**Pull request by samnordmann**
Adds support for `MatmulOp` in `HostIrExecutor`

The implementation makes it very easy to support new ops that `ExpressionEvaluator` can handle.

Stacked on top of 
- #2533
- #2535

### CODE REVIEW

**Comment by wujingyue** [comment]
> Stacked on top of

I'm curious what you've been using to stack PRs and how you like it :)

File: tests/cpp/test_host_irs.cpp [hunk 1]
```diff
@@ -708,6 +708,43 @@ INSTANTIATE_TEST_SUITE_P(
...
+  constexpr int64_t M = 64;
+  constexpr int64_t K = 128;
+  constexpr int64_t N = 256;
+
+  auto hic = std::make_unique<HostIrContainer>();
+  FusionGuard fg(hic.get());
+
+  TensorView* A = makeContigTensor(3);
+  TensorView* B = makeContigTensor(3);
+  TensorView* C = matmul(A, B);
```
Comment (wujingyue) [inline tests/cpp/test_host_irs.cpp:724]:
```suggestion
  TensorView* a = makeContigTensor(3);
  TensorView* b = makeContigTensor(3);
  TensorView* c = matmul(a, b);
```
makes styles consistent. https://google.github.io/styleguide/cppguide.html#Variable_Names

File: tests/cpp/test_host_irs.cpp [hunk 2]
```diff
@@ -708,6 +708,43 @@ INSTANTIATE_TEST_SUITE_P(
       return ss.str();
     });
 
+using MatmulHostIrTest = NVFuserTest;
+
+TEST_F(MatmulHostIrTest, HostIr) {
```
Comment (wujingyue) [inline tests/cpp/test_host_irs.cpp:713]:
```suggestion
TEST_F(HostIrTest, Matmul) {
```
I understand HostIrTest is a parameterized test, but it's fine to use it as a TEST_F as long as we don't call GetParam.

File: tests/cpp/test_host_irs.cpp [hunk 3]
```diff
@@ -708,6 +708,43 @@ INSTANTIATE_TEST_SUITE_P(
...
+  hic->addInput(A);
+  hic->addInput(B);
+  hic->addOutput(C);
+
+  hic->pushBackTopLevelExprs(C->definition());
+
+  HostIrExecutor hie(std::move(hic));
+
+  auto options = at::TensorOptions().device(at::kCUDA, 0).dtype(torch::kFloat);
+  c10::IValue A_ivalue = at::randn({H, M, K}, options);
```
Comment (wujingyue) [inline tests/cpp/test_host_irs.cpp]: I pushed a commit to make this simply an at::Tensor.

**Review by wujingyue — APPROVED** [review]

**Comment by samnordmann** [comment]
> > Stacked on top of
> 
> I'm curious what you've been using to stack PRs and how you like it :)

Nothing! Just manual management! :') I'd be happy to learn better ways though

File: tests/cpp/test_host_irs.cpp (hunk 2 above)
Comment (samnordmann) [reply tests/cpp/test_host_irs.cpp:713]: I tried that previously but had an error: https://github.com/NVIDIA/Fuser/pull/2491#discussion_r1660893569

Comment (samnordmann) [reply tests/cpp/test_host_irs.cpp:713]: let me know if I'm missing something

**Pull request merged**